src/ewatercycle/esmvaltool/search.py

```diff
@@ -0,0 +1,219 @@
+        in a set as items.
+    """
+    if CFG["search_esgf"] != "always":
+        msg = (
+            "The ESMValTool configuration option 'search_esgf' is not set to 'always'\n"
+            "this can lead to retrieving only subsets of the data you are interested "
+            "in.\n"
+            "To set the configuration to always search on ESGF, do:\n"
+            "    from esmvalcore.config import CFG\n"
+            "    CFG['search_esgf'] = 'always'\n"
+        )
+        warnings.warn(msg, category=UserWarning)
+
+    datasets = _query_esgf(
+        activity=activity, experiment=experiment, variables=variables, verbose=verbose
+    )
+
+    mips = _get_mip_tables(freq=frequency, extended=extended_mip_tables)
+    datasets = _filter_datasets(datasets, "mip", mips)
+
+    dataset_names = set([dataset["dataset"] for dataset in datasets])
+    unique_ensemble_members = set([dataset["ensemble"] for dataset in datasets])
+
+    valid_datasets: list[Dataset] = []
+    # Iterate through every dataset (i.e. model)
+    for dataset_name in dataset_names:
+        selected_dataset = _filter_datasets(datasets, "dataset", str(dataset_name))
+        # Iterate through every ensemble of this dataset
+        for ensemble in unique_ensemble_members:
+            selected_ensemble = _filter_datasets(
+                selected_dataset, "ensemble", str(ensemble)
+            )
+            output_short_names = _get_value(selected_ensemble, "short_name")
+
+            # Check if this model ensemble has all required variables
+            if all([var in output_short_names for var in variables]):
+                valid_datasets.extend(selected_ensemble)
+
+    valid_dataset_names = _get_value(valid_datasets, "dataset")
+    valid_ensembles: dict[str, set[str]] = dict()
+    for name in valid_dataset_names:
+        ensembles = _get_value(
+            _filter_datasets(valid_datasets, "dataset", str(name)),
+            "ensemble",
+        )
+        valid_ensembles[str(name)] = set(str(ens) for ens in ensembles)
+    return valid_ensembles
+
+
+def _query_esgf(
+    activity: str,
+    experiment: str,
+    variables: list[str],
+    verbose: bool = False,
+) -> list[Dataset]:
+    """Return all datasets on ESGF that match the specified search query.
+
+    Args:
+        activity: Activity type, for example: 'ScenarioMIP'.
+        experiment: Experiment within the activity. E.g.: 'ssp585'
+        variables: Which variables are you searching for. Use the short_name definition.
+            For example: ['pr', 'tas'].
+        verbose (optional): If the results should be printed in a verbose way, to aid
+            in your search experience. Defaults to False.
+
+    Returns:
+        List of ESMValTool 'Dataset' objects that match the search query.
+    """
+    datasets = list()
+
+    for var in variables:
+        dataset_query = Dataset(
+            short_name=var,
+            activity=activity,
+            mip="*",
```

Comment:
No, the mip facet here can only be a string, no list/tuple.

Iterating through both the var and mip also does not work, this will result in an error such as:
```
RecipeError: Unable to load CMOR table (project) 'CMIP6' for variable 'pr' with mip 'Eday'
```
Because you have to specify the correct mip per variable.